### PR TITLE
feat(billing): rework usage tracking + context breakdown

### PR DESCRIPTION
**Pull request by k11kirky**
## Problem

Users have no visibility into how much of their LLM usage quota they've consumed until they hit a hard limit, and the existing polling-based usage fetch was inefficient and prone to drift. Free-tier users lacked a reset-time label on the sidebar bar, and the context window indicator showed only a tooltip with aggregate numbers rather than a breakdown by source.

## Changes

**Usage monitoring service (`UsageMonitorService`)**
- Introduced a new `UsageMonitorService` in the main process that replaces the renderer's 30-second polling loop. The service listens for `LlmActivity` events emitted once per completed agent turn (both Claude and Codex adapters) and coalesces bursts into a single trailing fetch per 5-second window. A 30-minute backstop timer handles idle periods and billing-period rollovers.
- Threshold crossing detection fires `ThresholdCrossed` events at 50/75/90/100% for both `burst` and `sustained` buckets. Crossed thresholds are deduplicated per user/product/bucket/billing-window anchor and persisted to `electron-store` so notifications don't re-fire after relaunch within the same window. Stale entries are pruned on boot.
- Added `reset_at` (absolute UTC timestamp) and `billing_period_end` to the gateway schemas so the anchor logic doesn't drift with rolling `resets_in_seconds` values.

**tRPC surface**
- Replaced `llmGateway.usage` query with a dedicated `usageMonitor` router exposing `getLatest`, `refresh`, `onUsageUpdated` (subscription), and `onThresholdCrossed` (subscription).
- `useUsage` now subscribes to `onUsageUpdated` and seeds the query cache from `getLatest` instead of polling on a timer.

**Threshold toast notifications**
- `initializeUsageThresholdToast` subscribes to `onThresholdCrossed` and shows a warning toast at 50/75/90% with a reset-time label and a "View usage" action, or triggers the `UsageLimitModal` at 100% when a session is active.
- Removed the old `useUsageLimitDetection` hook.

**Sidebar usage bar**
- Shows a loading skeleton while data is in flight.
- Displays a human-readable reset-time label (`Resets in 4h 30m`, `Resets Jun 1 at 12:00 AM PDT`, etc.) derived from the new `formatResetTime` utility, which prefers `reset_at` over `resets_in_seconds`.

**Context breakdown popover**
- Added a `ContextBreakdownPopover` that replaces the plain tooltip on the `ContextUsageIndicator`. It shows a segmented bar and per-category token counts (System prompt, Tools, Rules, Skills, MCP, Subagents, Conversation).
- The breakdown is computed agent-side using a character-ratio estimator (~3.5 chars/token) and emitted via `_posthog/usage_update` notifications. Claude estimates system prompt, CLAUDE.md rules, slash-command skills, and MCP tool metadata at session init and on changes; Codex uses a constant baseline plus the injected system prompt.
- Extracted `formatTokensCompact`, `getOverallUsageColor`, and `CONTEXT_CATEGORIES` into a shared `contextColors.ts` utility.

**Codex session-state fix**
- Replaced `this.sessionState = createSessionState(...)` reassignments with a new `resetSessionState()` mutation so the codex-client closure always writes `contextUsed`/`contextSize` to the same object reference across `newSession`/`loadSession`/`resumeSession`/`forkSession` calls.

## How did you test this?

- Added unit tests for `UsageMonitorService` covering threshold deduplication, cross-relaunch persistence, independent burst/sustained tracking, `isPro` detection, error resilience, `UsageUpdated` change detection, coalesce debouncing, and `stop()` cleanup.
- Added unit tests for `context-breakdown.ts` covering all estimator functions and `buildBreakdown` edge cases.
- Added unit tests for `useContextUsage` covering aggregate extraction, breakdown merging, and the double-underscore method prefix variant.
- Added unit tests for `ContextBreakdownPopover` covering header rendering, placeholder copy, and per-category row filtering.
- Added a regression test for `createCodexClient` verifying that `contextUsed` writes land on the correct object after `resetSessionState`.
- Added unit tests for `formatResetTime` covering sub-hour, sub-day, multi-day, `reset_at` preference, and past-reset cases.

## Publish to changelog?

no

### CODE REVIEW

**Comment by k11kirky** [comment]
* **#2373** <a href="https://app.graphite.com/github/pr/PostHog/code/2373?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/PostHog/code/2373?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: apps/code/src/main/services/agent/schemas.ts [hunk 1]
```diff
@@ -203,6 +203,9 @@ export const AgentServiceEvent = {
   SessionsIdle: "sessions-idle",
   SessionIdleKilled: "session-idle-killed",
   AgentFileActivity: "agent-file-activity",
+  // Fires once per completed turn for both adapters. Consumed by
+  // UsageMonitorService to refresh billing usage without polling.
```
Comment (jonathanlab) [inline apps/code/src/main/services/agent/schemas.ts]: bit of a slopcomment

File: apps/code/src/main/services/llm-gateway/schemas.ts [hunk 2]
```diff
@@ -60,6 +60,9 @@ export interface AnthropicErrorResponse {
 export const usageBucketSchema = z.object({
   used_percent: z.number(),
   resets_in_seconds: z.number(),
+  // Absolute UTC reset timestamp from gateway A1; preferred over the
```
Comment (jonathanlab) [inline apps/code/src/main/services/llm-gateway/schemas.ts]: what's gateway A1..?

File: apps/code/src/main/services/usage-monitor/service.ts [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+  }
+
+  private processUsage(usage: UsageOutput): void {
+    const userId = usage.user_id.toString();
+    const product = usage.product;
+    // Plan-key isn't on UsageOutput; the only signal we have client-side is
+    // whether limits are at the Pro tier — but fetchUsage doesn't return that
+    // either. Best-effort: assume Pro if billing_period_end is present
+    // (free users never have it).
+    const isPro = !!usage.billing_period_end;
```
Comment (jonathanlab) [inline apps/code/src/main/services/usage-monitor/service.ts]: This seems very hacky, is this the best way to go about this?

File: apps/code/src/main/trpc/routers/usage-monitor.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    const iterable = service.toIterable(event, { signal: opts.signal });
+    for await (const data of iterable) {
+      yield data;
+    }
+  });
+}
+
+export const usageMonitorRouter = router({
+  onThresholdCrossed: subscribe(UsageMonitorEvent.ThresholdCrossed),
+  // Stream of full usage snapshots — replaces the renderer's 30s poll.
```
Comment (jonathanlab) [inline apps/code/src/main/trpc/routers/usage-monitor.ts]: There's a lot of comments referring what the code used to be like, such as

>  // Stream of full usage snapshots — replaces the renderer's 30s poll.

Lets get rid of those, they dont add a lot

File: packages/agent/src/adapters/claude/context-breakdown.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+  | "conversation";
+
+export type ContextBreakdown = Record<ContextCategory, number>;
+
+// Rough estimate of Claude's bundled `claude_code` preset system prompt. The
+// preset content is opaque to us so we add this constant when the systemPrompt
+// uses the preset — otherwise it'd show up as Conversation and skew the chart.
+const CLAUDE_PRESET_ESTIMATE_TOKENS = 4000;
+
+const CHARS_PER_TOKEN = 3.5;
```
Comment (jonathanlab) [inline packages/agent/src/adapters/claude/context-breakdown.ts]: This should be 4 according to Anthropic: 
>Tokens are pieces of text that models process. As a rough estimate, 1 token is approximately 4 characters or 0.75 words in English. The exact count varies by language and content type.

File: apps/code/src/main/services/usage-monitor/service.ts [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+    try {
+      return await this.llmGateway.fetchUsage();
+    } catch (err) {
+      log.debug("Usage fetch skipped", {
+        error: err instanceof Error ? err.message : String(err),
+      });
+      return null;
+    }
+  }
+
```
Comment (jonathanlab) [inline apps/code/src/main/services/usage-monitor/service.ts]: Why do we need this?

File: apps/code/src/renderer/App.tsx [hunk 7]
```diff
@@ -63,6 +64,12 @@ function App() {
     };
   }, []);
 
+  // Initialize usage threshold notifications (50/75/90/100%)
+  useEffect(() => {
+    if (!isAuthenticated) return;
+    return initializeUsageThresholdToast();
+  }, [isAuthenticated]);
+
```
Comment (jonathanlab) [inline apps/code/src/renderer/App.tsx]: See R9 of the new architecture guidelines we just merged in, maybe just run this PR against those freshly merged rules for general architectural improvements

**Review by jonathanlab — COMMENTED** [review]
would recommend running this agents the newly merged `AGENTS.md`

File: apps/code/src/main/services/usage-monitor/service.ts (hunk 6 above)
Comment (k11kirky) [reply apps/code/src/main/services/usage-monitor/service.ts]: for anything that is a change outside of an llm activity. i.e plan upgrades from the dashboard, admin quota grants, we need the poll to protect against these

**Review by jonathanlab — APPROVED** [review]
